src/coreclr/jit/optimizer.cpp

```diff
@@ -4662,267 +5010,311 @@ PhaseStatus Compiler::optUnrollLoops()
+// +--*  CNS_INT   int    9 $43
+//
+// becomes:
+//
+// ------------ BB03 [009..00D) -> BB05 (always), preds={BB02} succs={BB05}
+// STMT00004
+//   *  NOP       void
+//
+// STMT00005
+//   *  RETURN    int    $VN.Void
+//   \--*  SELECT    int
+//      +--*  LT        int    $102
+//      |  +--*  LCL_VAR   int    V02
+//      |  \--*  CNS_INT   int    7 $46
+//      +--*  CNS_INT   int    5 $41
+//      +--*  CNS_INT   int    9 $43
+//
+// STMT00006
+//   *  NOP       void
+//
+// ------------ BB04 [00D..010), preds={} succs={BB06}
+// ------------ BB05 [00D..010), preds={} succs={BB06}
+//
+bool OptIfConversionDsc::optIfConvert()
 {
 #ifndef TARGET_ARM64
     return false;
-#else
+#endif
 
     // Don't optimise the block if it is inside a loop
     // When inside a loop, branches are quicker than selects.
     // Detect via the block weight as that will be high when inside a loop.
-    if ((block->getBBWeight(this) > BB_UNITY_WEIGHT) && !compStressCompile(STRESS_IF_CONVERSION_INNER_LOOPS, 25))
+    if ((m_startBlock->getBBWeight(m_comp) > BB_UNITY_WEIGHT) &&
+        !m_comp->compStressCompile(m_comp->STRESS_IF_CONVERSION_INNER_LOOPS, 25))
     {
         return false;
     }
 
     // Does the block end by branching via a JTRUE after a compare?
-    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    if (m_startBlock->bbJumpKind != BBJ_COND || m_startBlock->NumSucc() != 2)
     {
         return false;
     }
 
     // Verify the test block ends with a condition that we can manipulate.
-    GenTree* last = block->lastStmt()->GetRootNode();
+    GenTree* last = m_startBlock->lastStmt()->GetRootNode();
     noway_assert(last->OperIs(GT_JTRUE));
     GenTree* cond = last->gtGetOp1();
     if (!cond->OperIsCompare())
     {
         return false;
     }
-
-    // Block where the flows merge.
-    BasicBlock* finalBlock = block->bbNext;
-    // The node, statement and block of the assignment.
-    GenTree*    asgNode  = nullptr;
-    Statement*  asgStmt  = nullptr;
-    BasicBlock* asgBlock = nullptr;
-
-    // Check the block is followed by a block or chain of blocks that only contain NOPs and
-    // a single ASG statement. The destination of the final block must point to the same as the
-    // true path of the JTRUE block.
-    bool foundMiddle = false;
-    while (!foundMiddle)
+    // Check the condition or inverted condition is usable. Prefer inverted.
+    bool         invertCond = true;
+    GenCondition genCond    = GenCondition::FromRelop(cond);
+    if (!GenCondition::IsPreferredRelop(GenCondition::Reverse(genCond)))
     {
-        BasicBlock* middleBlock = finalBlock;
-        noway_assert(middleBlock != nullptr);
-
-        // middleBlock should have a single successor.
-        finalBlock = middleBlock->GetUniqueSucc();
-        if (finalBlock == nullptr)
+        invertCond = false;
+        if (!GenCondition::IsPreferredRelop(genCond))
         {
             return false;
         }
+    }
```

Comment:
> For Select I was going to say it's a lot harder, as it'll require additional register to merge flags. But after drawing it out, I think we can do it via two csel instructions and no extra registers.
> 
> I'll redo this part and make sure there's an explicit test for it too.

I don't think you necessarily have to do this work in this PR. What I mean is that the JIT should already be able to handle these "multi-jump" conditional ops when they are being consumed as a value. Something like
```csharp
float a = ..., b = ...;
bool c = a != b;
```
will already produce something that the parent node does not need special knowledge to handle. `GT_SELECT` can work in the same way, simply by avoiding containing the compare in the cases that it does not (yet) handle. It will produce less efficent code than your suggested approach, but I think leaving it to a future PR is fine (or skipping if-conversion for float ops we cannot handle for now).

Reply:
Worth noting that this code is currently wrong in HEAD.  Not quite sure yet why nothing has hit yet. Might be worth splitting this part into a separate PR to get it merged quicker.

Reply:
Sounds fine to me.

Reply:
This part is in a new PR: https://github.com/dotnet/runtime/pull/78223